lib/util/zip.js

```diff
@@ -47,29 +47,35 @@ const zipStreamFromParts = (...zipParts) => {
     if (cb != null) cb();
   });
 
-  for (const part of zipParts) {
-    part.stream.on('data', ({ stream, options, cb }) => {
-      const s = (stream instanceof PartialPipe)
-        ? stream.pipeline((err) => { resultStream.emit('error', err); resultStream.abort(); })
-        : stream;
+  const next = () => {
+    if (zipPartFns.length === 0) {
+      resultStream.finalize();
+      return;
+    }
 
-      if (cb == null) {
-        resultStream.append(s, options);
-      } else {
-        // using the String object will result in still an empty comment, but allows
-        // separate instance equality check when the entry is recorded.
-        const sentinel = new String(); // eslint-disable-line no-new-wrappers
-        callbacks.set(sentinel, cb);
-        resultStream.append(s, Object.assign(options, { comment: sentinel }));
-      }
-    });
-    part.stream.on('error', (err) => { resultStream.emit('error', err); });
-    part.stream.on('end', () => { // eslint-disable-line no-loop-func
-      completed += 1;
-      if (completed === zipParts.length)
-        resultStream.finalize();
-    });
-  }
+    resolve(zipPartFns.shift()())
```

Comment:
This is the main difference from the approach in #564. In this PR, `stream()` returns a Promise of a stream, so `zipStreamFromParts()` needs some Promise-related logic. However, in #564, `stream()` simply returns a stream.